From 4a9d0ed98a644c751536a7a0308fb76eb9338420 Mon Sep 17 00:00:00 2001
From: Gatien Chevallier <gatien.chevallier@foss.st.com>
Date: Mon, 2 Oct 2023 17:02:38 +0200
Subject: [PATCH 362/841] feat(st-drivers): add RIFSC driver

Add the RIFSC new driver support.

RIFSC (RIF Security Controller) is responsible for the isolation
of hardware resources like memory or peripherals. It is composed of:

-RISC registers(slave peripherals) with RISUP(Resource Isolation
Slave Unit for Peripherals) OR RISAL(Resource Isolation Slave Unit
for Address space - Lite) logics.
-RIMC registers(Non RIF-Aware masters counterpart) with RIMU
(Resource Isolation Master Unit) logic. It is possible for a master to
inherit from its slave port(RISUP) configuration.

This driver only supports semaphore acquisition for some IDs.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/arm-trusted-firmware.git 4b2e31fc2bf733c0816ab9fa9990d45366868267 feat(st-drivers): add RIFSC driver]
Signed-off-by: Gatien Chevallier <gatien.chevallier@foss.st.com>
Change-Id: Iba4cdbf53243292fa0b42cad8392c43734dd9bc2
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/tf-a/+/331018
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Reviewed-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
Domain-Review: Yann GAUTIER <yann.gautier@foss.st.com>
Tested-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
Reviewed-by: Yann GAUTIER <yann.gautier@foss.st.com>
---
 drivers/st/rif/stm32_rifsc.c      | 68 +++++++++++++++++++++++++++++++
 include/drivers/st/stm32_rifsc.h  | 17 ++++++++
 plat/st/stm32mp2/bl2_plat_setup.c |  5 +++
 plat/st/stm32mp2/platform.mk      |  4 +-
 plat/st/stm32mp2/stm32mp2_def.h   |  3 ++
 5 files changed, 96 insertions(+), 1 deletion(-)
 create mode 100644 drivers/st/rif/stm32_rifsc.c
 create mode 100644 include/drivers/st/stm32_rifsc.h

diff --git a/drivers/st/rif/stm32_rifsc.c b/drivers/st/rif/stm32_rifsc.c
new file mode 100644
index 000000000..b0430bb09
--- /dev/null
+++ b/drivers/st/rif/stm32_rifsc.c
@@ -0,0 +1,68 @@
+/*
+ * Copyright (c) 2023, STMicroelectronics - All Rights Reserved
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#include <errno.h>
+#include <stdint.h>
+
+#include <arch_helpers.h>
+#include <drivers/st/stm32_rifsc.h>
+#include <dt-bindings/soc/rif.h>
+#include <lib/mmio.h>
+
+#include <platform_def.h>
+
+/* RIFSC general register field description */
+/* RIFSC_CIDCFGR register fields */
+#define _RIFSC_RISC_CIDCFGR(x)		(U(0x100) + U(0x8) * (x))
+#define _RIFSC_RISC_CFEN_MASK		BIT(0)
+#define _RIFSC_RISC_SEM_EN_MASK		BIT(1)
+#define _RIFSC_RISC_SEML_SHIFT		U(16)
+#define _RIFSC_RISC_SEML_MASK		GENMASK_32(23, 16)
+
+/* RIFSC_SEMCR register fields */
+#define _RIFSC_RISC_SEMCR(x)		(U(0x104) + U(0x8) * (x))
+#define _RIFSC_RISC_SEM_MUTEX		BIT(0)
+#define _RIFSC_RISC_SEMCID_SHIFT	U(4)
+#define _RIFSC_RISC_SEMCID_MASK		GENMASK_32(6, 4)
+
+static unsigned long rifsc_semaphores[] = {
+	STM32MP25_RIFSC_RNG_ID,
+	STM32MP25_RIFSC_PKA_ID,
+	STM32MP25_RIFSC_SAES_ID,
+};
+
+int stm32_rifsc_semaphore_init(void)
+{
+	unsigned long i = 0;
+
+	for (i = 0; i < ARRAY_SIZE(rifsc_semaphores); i++) {
+		uint32_t cidcfgr = mmio_read_32(RIFSC_BASE +
+						_RIFSC_RISC_CIDCFGR(rifsc_semaphores[i]));
+		uint32_t semcfgr = mmio_read_32(RIFSC_BASE +
+						_RIFSC_RISC_SEMCR(rifsc_semaphores[i]));
+
+		if (!(((cidcfgr & _RIFSC_RISC_CFEN_MASK) != 0U) &&
+		      ((cidcfgr & _RIFSC_RISC_SEM_EN_MASK) != 0U) &&
+		      ((cidcfgr & _RIFSC_RISC_SEML_MASK) >> _RIFSC_RISC_SEML_SHIFT) != RIF_CID1)) {
+			continue;
+		}
+
+		if (((semcfgr & _RIFSC_RISC_SEM_MUTEX) != 0U) &&
+		    ((semcfgr & _RIFSC_RISC_SEMCID_MASK) >> _RIFSC_RISC_SEMCID_SHIFT) != RIF_CID1) {
+			return -EACCES;
+		}
+
+		mmio_write_32(RIFSC_BASE + _RIFSC_RISC_SEMCR(rifsc_semaphores[i]),
+			      _RIFSC_RISC_SEM_MUTEX);
+
+		if (((semcfgr & _RIFSC_RISC_SEM_MUTEX) != 0U) &&
+		    ((semcfgr & _RIFSC_RISC_SEMCID_MASK) >> _RIFSC_RISC_SEMCID_SHIFT) != RIF_CID1) {
+			return -EACCES;
+		}
+	}
+
+	return 0;
+}
diff --git a/include/drivers/st/stm32_rifsc.h b/include/drivers/st/stm32_rifsc.h
new file mode 100644
index 000000000..b720eac98
--- /dev/null
+++ b/include/drivers/st/stm32_rifsc.h
@@ -0,0 +1,17 @@
+/*
+ * Copyright (c) 2023, STMicroelectronics - All Rights Reserved
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#ifndef STM32MP2_RIFSC_H
+#define STM32MP2_RIFSC_H
+
+/*
+ * stm32_rifsc_semaphore_init() - Takes RIFSC semaphore for some IDs
+ *
+ * Return 0 on success and -EACCES if authorized semaphore couldn't be taken
+ */
+int stm32_rifsc_semaphore_init(void);
+
+#endif /* STM32MP2_RIFSC_H */
diff --git a/plat/st/stm32mp2/bl2_plat_setup.c b/plat/st/stm32mp2/bl2_plat_setup.c
index 3c17523ea..60f79446a 100644
--- a/plat/st/stm32mp2/bl2_plat_setup.c
+++ b/plat/st/stm32mp2/bl2_plat_setup.c
@@ -21,6 +21,7 @@
 #include <drivers/st/regulator_fixed.h>
 #include <drivers/st/stm32_console.h>
 #include <drivers/st/stm32_iwdg.h>
+#include <drivers/st/stm32_rifsc.h>
 #include <drivers/st/stm32_rng.h>
 #include <drivers/st/stm32mp_pmic2.h>
 #include <drivers/st/stm32mp_reset.h>
@@ -275,6 +276,10 @@ skip_console_init:
 	}
 #endif
 
+	if (stm32_rifsc_semaphore_init() != 0) {
+		panic();
+	}
+
 	if (stm32_rng_init() != 0) {
 		panic();
 	}
diff --git a/plat/st/stm32mp2/platform.mk b/plat/st/stm32mp2/platform.mk
index a5afc3add..3a4f13df3 100644
--- a/plat/st/stm32mp2/platform.mk
+++ b/plat/st/stm32mp2/platform.mk
@@ -191,7 +191,9 @@ BL2_SOURCES			+=	drivers/st/crypto/stm32_hash.c				\
 					plat/st/stm32mp2/bl2_plat_setup.c			\
 					plat/st/stm32mp2/plat_ddr.c
 
-BL2_SOURCES		+=	drivers/st/rif/stm32mp2_risaf.c
+BL2_SOURCES		+=	drivers/st/rif/stm32_rifsc.c \
+				drivers/st/rif/stm32mp2_risaf.c
+
 
 ifeq (${TRUSTED_BOARD_BOOT},1)
 BL2_SOURCES		+=	drivers/st/crypto/stm32_pka.c
diff --git a/plat/st/stm32mp2/stm32mp2_def.h b/plat/st/stm32mp2/stm32mp2_def.h
index b91441788..206703e89 100644
--- a/plat/st/stm32mp2/stm32mp2_def.h
+++ b/plat/st/stm32mp2/stm32mp2_def.h
@@ -507,6 +507,9 @@ static inline uintptr_t tamp_bkpr(uint32_t idx)
  * RIFSC IDs
  ******************************************************************************/
 #define STM32MP25_RIFSC_USB3DR_ID	U(66)
+#define STM32MP25_RIFSC_RNG_ID		U(92)
+#define STM32MP25_RIFSC_PKA_ID		U(93)
+#define STM32MP25_RIFSC_SAES_ID		U(94)
 
 #define STM32MP25_RIMU_USB3DR		U(4)
 
-- 
2.47.2

